File: 200-number-of-islands/200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        visit = set()
        count = 0
        
        def dfs(r, c):
            if r in [-1, ROWS] or c in [-1, COLS] or (r, c) in visit or grid[r][c] == "0":
                return
            visit.add((r, c))
            for dr, dc in directions:
                dfs(r + dr, c + dc)
            
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == "1" and (r, c) not in visit:
                    count += 1
                    dfs(r, c)
                    
        return count

File: 200-number-of-islands/200-number-of-islands.java
class Solution {
    private static final int[][] DIRECTIONS = new int[][] {
        new int[] { 1, 0 },
        new int[] { -1, 0 },
        new int[] { 0, 1 },
        new int[] { 0, -1 }
    };
    
    public int numIslands(char[][] grid) {
        int islands = 0;
        for (int r = 0; r < grid.length; r++) {
            for (int c = 0; c < grid[0].length; c++) {
                if (grid[r][c] == '1') {
                    dfs(grid, r, c);
                    islands++;
                }
            }
        }
        return islands;
    }
    
    private void dfs(char[][] grid, int r, int c) {
        if (isOutOfBounds(grid, r, c) || grid[r][c] == '0') {
            return;
        }
        
        grid[r][c] = '0';
        
        for (int[] dir : DIRECTIONS) {
            dfs(grid, r + dir[0], c + dir[1]);
        }
    }
    
    private boolean isOutOfBounds(char[][] grid, int r, int c) {
        return r < 0 || r >= grid.length || c < 0 || c >= grid[0].length;
    }
}